### DISCUSSION TITLE
Can i use the Web API to export/import the schema of a pocketbase instance?

### DISCUSSION THREAD

**Discussion by snorresovold** [pocketbase/pocketbase#1999] [Q&A]
i want to code a script that imports/exports the schema of my pocketbase instances, is this possible?

**Comment by ganigeorgiev** [comment]
As mentioned earlier in https://github.com/pocketbase/pocketbase/discussions/1995#discussioncomment-5226086, you can use the [Collections web APIs](https://pocketbase.io/docs/api-collections/), for example:

```js
// authorize as admin
await pb.admins.authWithPassword('test@example.com', '1234567890');

// export
const collections = await pb.collections.getFullList();

// import
await pb.collections.import(collections, true);
```

**Reply by snorresovold** [reply]
I keep getting Invalid collections configuration when using input from 
`const collections = await pb.collections.getFullList();`

**Reply by snorresovold** [reply]
Nvm fixed it, I had an unrelated error that messed up the structure of the schema

**Reply by ganigeorgiev** [reply]
A lot of things could result to this error message (duplicated fields, system collection/fields removal, etc.) and it is hard to say what could be the reason for the failure.

You can start the executable with `--debug` and/or check the more detailed error message in the "Admin UI > Logs".

If the above doesn't help you pinpoint the issue, please share the collections configuration JSONs of the 2 environments you want to sync, so that I can try to reproduce it locally.

**Reply by snorresovold** [reply]
No it was a stupid mistake, I had written 
`const collections = await pb.collections.getFullList({ sort: '-created' });`

instead of 

`const collections = await pb.collections.getFullList();`

I didn't realise I didn't have to use an argument.

**Reply by ganigeorgiev** [reply]
Hm, the order of the collections inside the array shouldn't matter since they are inserted within a transaction directly and the schema validation happens after the insert, so I doubt that the `sort` parameter is the culprit here. You can check the logs for a more detailed error message.

Also keep in mind that the `getFullList` arguments changed recently with [JS SDK v0.11.0](https://github.com/pocketbase/js-sdk/releases/tag/v0.11.0) and if you are using an older version, the first argument must be the batch size.

**Reply by snorresovold** [reply]
Yes, I got my script fixed and working. This might be unrelated but how should I share this with the community?

It syncronises the schemas of 2 pocketbase instances, but I don't know what to do with it. Should i make an npm package or just add the code (it is a simple function) to something like [Awesome_pocketbase](https://github.com/benallfree/awesome-pocketbase).

**Reply by ganigeorgiev** [reply]
> This might be unrelated but how should I share this with the community?

It is up to you. I still don't understand what is your use case and why do you prefer to manually import the schemas with the API (_keep in mind that you may not always be able to do it since conflicts could arise if the 2 environments do changes independently; this is why the Admin UI Import interface was added_).

In general, the recommended approach for "transferring" collections changes from one environment to another is using [migrations](https://pocketbase.io/docs/migrations/) (aka. the `pb_migrations` dir for the prebuilt executable) because it is seamless and the autogenerated changes are granular allowing multiple developers to work on the same project without worrying much of overwriting the other dev(s) changes.

**Reply by snorresovold** [reply]
Well, I wanted to automate the schema migrations, and I have a School project about devops where I thought i would do this as well as some other stuff. I do still think this could be useful however as a devops tool and used in stuff like Github actions as an automation tool.

**Reply by ganigeorgiev** [reply]
But they are already automated. The only requirement is to copy you `pb_migrations` dir to the other environment. All unapplied migrations are automatically executed on `serve`. Similar to `pb_public`, this directory is safe to be committed in your repo.

In any case, I'm not saying that using the import API is wrong - sometimes it may be the only option that you have, but it has limitations and conflicts during the import process could arise. This is why the Admin UI Import interface was added to allow developers who want to import a JSON schema to review the changes manually and decide whether they need to adjust it (_eg. changing the collection and field ids_) or to try to import it as it is.

**Reply by snorresovold** [reply]
Yes, they are very automated "under the hood", but I'm looking for Full automation, i want to send a pull request and let github workflows do the rest. No need to log in. I know this isn't the way it's meant to be done but this script isn't meant to be used on anything "important". 

In any case, I learned a lot from it, and did what i wanted to to. This is all just meant to be for a school project about automation and devops, and all i need to do now is set it up with github workflows.
